src/browser/connection/index.js

```diff
@@ -112,9 +112,30 @@ export default class BrowserConnection extends EventEmitter {
         }
     }
 
+    _createBrowserDisconnectedError () {
+        return new GeneralError(MESSAGE.browserDisconnected, this.userAgent);
+    }
+
+    _onBrowserDisconnected () {
+        this.emit('error', this._createBrowserDisconnectedError());
+    }
+
+    _restartBrowser () {
+        this.ready = false;
+
+        this._forceIdle();
+
+        this._closeBrowser()
+            .then(() => this._runBrowser());
+    }
+
     _waitForHeartbeat () {
         this.heartbeatTimeout = setTimeout(() => {
-            this.emit('error', new GeneralError(MESSAGE.browserDisconnected, this.userAgent));
+            this.opened = false;
+
+            if (!this.emit('disconnected', this._createBrowserDisconnectedError()))
```

Comment:
I feel that it's better to allow suppressing the `error` event emission. E.g.
```
class BrowserConnection {
    constructor(){
         this.errorSuppressed = false;
    }

    _waitForHearbeat () {
        setTimeout(() => {
            this.emit('disconnect');
       
            if (!this.errorSuppressed)
                //emit error event
        }, ...);
    }
}

//Somewhere in test-run-controller.js

connection.on('disconnect', () => {
    if (disconnectionCount<DISCONNECTION_THRESHOLD)
        connection.errorSuppressed = true;
    ...
})
```

Of course, it's better to introduce an additional public method (e.g. `suppressError()`) to set the `errorSuppressed` to `true`.